File: php/features.php
<?php
declare(strict_types=1);

// FirstNews SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class FirstNewsFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new FirstNewsBaseFeature();
            case "ratelimit":
                return new FirstNewsRatelimitFeature();
            case "retry":
                return new FirstNewsRetryFeature();
            case "test":
                return new FirstNewsTestFeature();
            case "timeout":
                return new FirstNewsTimeoutFeature();
            default:
                return new FirstNewsBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: rb/features.rb
# FirstNews SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module FirstNewsFeatures
  def self.make_feature(name)
    case name
    when "base"
      FirstNewsBaseFeature.new
    when "ratelimit"
      FirstNewsRatelimitFeature.new
    when "retry"
      FirstNewsRetryFeature.new
    when "test"
      FirstNewsTestFeature.new
    when "timeout"
      FirstNewsTimeoutFeature.new
    else
      FirstNewsBaseFeature.new
    end
  end
end
